cmd/zc_traverser_list.go

```diff
@@ -0,0 +1,127 @@
+package cmd
+
+import (
+	"bufio"
+	"fmt"
+	"github.com/Azure/azure-storage-azcopy/common"
+	"net/url"
+	"os"
+)
+
+// a meta traverser that goes through a list of paths (potentially directory entities) and scans them one by one
+// behaves like a single traverser (basically a "traverser of traverser")
+type listTraverser struct {
+	fileLocation            string
+	childTraverserGenerator childTraverserGenerator
+}
+
+type childTraverserGenerator func(childPath string) (resourceTraverser, error)
+
+func (l *listTraverser) traverse(processor objectProcessor, filters []objectFilter) (err error) {
+	f, err := os.Open(l.fileLocation)
+	if err != nil {
+		return fmt.Errorf("unable to open %s to retrieve the required list of entities to transfer", l.fileLocation)
+	}
+	defer f.Close()
+
+	// spawn a scanner to read the list of entities one line at a time
+	scanner := bufio.NewScanner(f)
+	for scanner.Scan() {
+		childPath := scanner.Text()
+
+		// fetch an appropriate traverser, and go through the child path, which could be
+		//   1. a single entity
+		//   2. a directory entity that needs to be scanned
+		childTraverser, err := l.childTraverserGenerator(childPath)
+		if err != nil {
+			glcm.Info(fmt.Sprintf("Skipping %s due to error %s", childPath, err))
+			continue
+		}
+
+		err = childTraverser.traverse(processor, filters)
+		if err != nil {
+			glcm.Info(fmt.Sprintf("Skipping %s as it cannot be scanned due to error: %s", childPath, err))
+		}
+	}
+
+	// in case the list of entities was not read properly, we can no longer continue enumeration
+	if err := scanner.Err(); err != nil {
+		return fmt.Errorf("unable to scan %s to retrieve the required list of entities to transfer", l.fileLocation)
+	}
+
+	return nil
+}
+
+func (l *listTraverser) setPreProcessingCallback(callback preProcessingCallback) {
```

Comment:
Is it possible we could just hand this off to all sub-traversers?

Reply:
(alternatively, execute the specified callback before we execute our specified callback)

Reply:
Yes, but the drawback is that filters see the incomplete relative path. I think we can live with it ok.